### PR TITLE
Remove `--api-port` flag from the cli

### LINKED ISSUE #2351: Remove --api-port from inject and install
The --api-port is superfluous for install and inject -- there's no good reason a user would want to configure it. We should remove it.

### PR DESCRIPTION
**Pull request by paranoidaditya**
Closes #2351

Signed-off-by: Aditya Sharma <hello@adi.run>

### CODE REVIEW

**Comment by paranoidaditya** [comment]
@olix0r Quick and easy, if you could take a 👀

**Comment by ihcsim** [comment]
@paranoidaditya Thanks for working on this. In addition to the CLI flag, it will be great if you can also remove the same variable from the `proxyConfigOptions` as seen in the CLI source and test:
```
$ grep -rniI destinationapiport cli
cli/cmd/install_test.go:39:             DestinationAPIPort:               123,
cli/cmd/install_test.go:92:             DestinationAPIPort:               123,
cli/cmd/inject.go:262:                          Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, options.destinationAPIPort),
cli/cmd/testdata/install_ha_with_overrides_output.golden:288:    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy:dev-undefined","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init:dev-undefined","pullPolicy":"IfNotPresent"},"destinationApiPort":{"port":8086},"controlPort":{"port":4190},"ignoreInboundPorts":[],"ignoreOutboundPorts":[],"inboundPort":{"port":4143},"metricsPort":{"port":4191},"outboundPort":{"port":4140},"resource":{"requestCpu":"400m","requestMemory":"300Mi","limitCpu":"","limitMemory":""},"proxyUid":"2102","logLevel":{"level":"warn,linkerd2_proxy=info"},"disableExternalProfiles":false}
cli/cmd/testdata/install_no_init_container_auto_inject.golden:282:    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy:dev-undefined","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init:dev-undefined","pullPolicy":"IfNotPresent"},"destinationApiPort":{"port":8086},"controlPort":{"port":4190},"ignoreInboundPorts":[],"ignoreOutboundPorts":[],"inboundPort":{"port":4143},"metricsPort":{"port":4191},"outboundPort":{"port":4140},"resource":{"requestCpu":"","requestMemory":"","limitCpu":"","limitMemory":""},"proxyUid":"2102","logLevel":{"level":"warn,linkerd2_proxy=info"},"disableExternalProfiles":false}
cli/cmd/testdata/install_ha_output.golden:288:    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy:dev-undefined","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init:dev-undefined","pullPolicy":"IfNotPresent"},"destinationApiPort":{"port":8086},"controlPort":{"port":4190},"ignoreInboundPorts":[],"ignoreOutboundPorts":[],"inboundPort":{"port":4143},"metricsPort":{"port":4191},"outboundPort":{"port":4140},"resource":{"requestCpu":"10m","requestMemory":"20Mi","limitCpu":"","limitMemory":""},"proxyUid":"2102","logLevel":{"level":"warn,linkerd2_proxy=info"},"disableExternalProfiles":false}
cli/cmd/testdata/install_no_init_container.golden:252:    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy:dev-undefined","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init:dev-undefined","pullPolicy":"IfNotPresent"},"destinationApiPort":{"port":8086},"controlPort":{"port":4190},"ignoreInboundPorts":[],"ignoreOutboundPorts":[],"inboundPort":{"port":4143},"metricsPort":{"port":4191},"outboundPort":{"port":4140},"resource":{"requestCpu":"","requestMemory":"","limitCpu":"","limitMemory":""},"proxyUid":"2102","logLevel":{"level":"warn,linkerd2_proxy=info"},"disableExternalProfiles":false}
cli/cmd/testdata/install_default.golden:276:    {"proxyImage":{"imageName":"gcr.io/linkerd-io/proxy:dev-undefined","pullPolicy":"IfNotPresent"},"proxyInitImage":{"imageName":"gcr.io/linkerd-io/proxy-init:dev-undefined","pullPolicy":"IfNotPresent"},"destinationApiPort":{"port":8086},"controlPort":{"port":4190},"ignoreInboundPorts":[],"ignoreOutboundPorts":[],"inboundPort":{"port":4143},"metricsPort":{"port":4191},"outboundPort":{"port":4140},"resource":{"requestCpu":"","requestMemory":"","limitCpu":"","limitMemory":""},"proxyUid":"2102","logLevel":{"level":"warn,linkerd2_proxy=info"},"disableExternalProfiles":false}
cli/cmd/root.go:258:    destinationAPIPort      uint
cli/cmd/root.go:294:            destinationAPIPort:      8086,
cli/cmd/root.go:400:    cmd.PersistentFlags().UintVar(&options.destinationAPIPort, "api-port", options.destinationAPIPort, "Port where the Linkerd controller's destination API is running")
cli/cmd/install.go:41:  DestinationAPIPort               uint
cli/cmd/install.go:217:         DestinationAPIPort:               options.destinationAPIPort,
cli/cmd/install.go:391:         DestinationApiPort: &config.Port{
cli/cmd/install.go:392:                 Port: uint32(options.destinationAPIPort),
```

Note that this variable is also used in the [`proto/config/config.proto`](https://github.com/linkerd/linkerd2/blob/21849288135eacf826497cbaf1faba5361b6ab5e/proto/config/config.proto#L21) file. But since changing this file will lead to conflicts with some ongoing PRs, you can leave it out in this PR. Thanks!

**Comment by ihcsim** [comment]
@paranoidaditya Thanks for addressing my feedback. There are some conflicts with a recent merge in `master`. I think the easiest thing to do is to create a new branch off latest `master`, add your changes there and then open a new PR. Per our convo yesterday, you can run `go test ./cli/... --update` to auto-fix all the broken unit tests. Thanks again!

**Comment by grampelberg** [comment]
I usually use `git cherry-pick` to do something like @ihcsim is outlining.

**Comment by paranoidaditya** [comment]
Rebased and pushed, should be good to go now.

I chatted with @ihcsim regarding the changing the protobuf file to take out `destinationApiPort` entirely and we agreed on holding off on that due to conflicts with ongoing PRs

**Comment by ihcsim** [comment]
@klingerf @siggy I wonder if we should take this one step further and remove the `destinationAPIPort` completely from the helm chart and the `destination` server's `-addr` flag? Presumably, we still need the `LINKERD2_PROXY_CONTROL_URL` env var in the proxy spec?

Otherwise, the rest LGTM, including not touching `config.proto` for now, since there are other changes in #2434.

**Comment by siggy** [comment]
@ihcsim Just curious, are the `"destinationApiPort":null` references in the test outputs an artifact of `destinationApiPort` still being part of `config.proto`? If so, and there's other changes coming in #2434, I agree it makes sense to leave it for now.

I agree we should leave `LINKERD2_PROXY_CONTROL_URL`, and similarly the `-addr` flag on the destination server. I am speculating that there may be debugging use cases where I would want to test multiple destinations servers locally, so changing the port is helpful.

File: pkg/inject/inject.go [hunk 1]
```diff
@@ -413,7 +413,7 @@ func (conf *ResourceConfig) injectPodSpec(patch *Patch, identity k8s.TLSIdentity
 			{Name: "LINKERD2_PROXY_LOG", Value: conf.proxyConfig.GetLogLevel().GetLevel()},
 			{
 				Name:  "LINKERD2_PROXY_CONTROL_URL",
-				Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, conf.proxyConfig.GetDestinationApiPort().GetPort()),
+				Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, 8086),
```
Comment (klingerf) [inline pkg/inject/inject.go]: Let's add a constant at the top of this file, near where `defaultKeepaliveMs` is defined, to hold this port value. I'm just a bit worried about it getting lost in the configuration code. Something like:

```diff
diff --git a/pkg/inject/inject.go b/pkg/inject/inject.go
index 938a6e20..9f40a0aa 100644
--- a/pkg/inject/inject.go
+++ b/pkg/inject/inject.go
@@ -31,6 +31,8 @@ const (
 	podNamespaceEnvVarName = "LINKERD2_PROXY_POD_NAMESPACE"
 	// defaultKeepaliveMs is used in the proxy configuration for remote connections
 	defaultKeepaliveMs = 10000
+	// destinationAPIPort is the port exposed by the linkerd-destination service
+	destinationAPIPort = 8086
 )
 
 var injectableKinds = []string{
@@ -420,7 +422,7 @@ func (conf *ResourceConfig) injectPodSpec(patch *Patch, identity k8s.TLSIdentity
 			{Name: "LINKERD2_PROXY_LOG", Value: conf.proxyConfig.GetLogLevel().GetLevel()},
 			{
 				Name:  "LINKERD2_PROXY_CONTROL_URL",
-				Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, 8086),
+				Value: fmt.Sprintf("tcp://%s:%d", controlPlaneDNS, destinationAPIPort),
 			},
 			{Name: "LINKERD2_PROXY_CONTROL_LISTENER", Value: fmt.Sprintf("tcp://0.0.0.0:%d", conf.proxyConfig.GetControlPort().GetPort())},
 			{Name: "LINKERD2_PROXY_METRICS_LISTENER", Value: fmt.Sprintf("tcp://0.0.0.0:%d", conf.proxyConfig.GetMetricsPort().GetPort())},
```

**Comment by ihcsim** [comment]
The `"destinationApiPort":null` is a result of the change in `cli/cmd/install.go` where the `DestinationApiPort` assignment is removed from the proxy config proto. (We instantiated the `jsonpb.Marshaler` with `EmitDefaults: true`.) My thought is that this field should be removed from the `config.proto` too (perhaps in #2434, or another PR) .

**Comment by paranoidaditya** [comment]
@klingerf Is renumbering protobuf fields okay or should it stay the same for backward compatibility?

File: pkg/inject/inject.go (hunk 1 above)
Comment (klingerf) [reply pkg/inject/inject.go]: I see this conversation marked as resolved, but I don't actually see the new constant in the diff. Is it possible this change didn't get pushed?

Comment (paranoidaditya) [reply pkg/inject/inject.go]: Oops, forgot to add the file.
Should be 👍  now

**Pull request merged**